requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,130 @@
+
+The multiple environments use case is one of the reasons why projects are a feature of PostHog, and projects are great for tracking separate products, but not ideal for tracking multiple instances of the same product: whatever you set up for one project, you have to manually replicate in all others.
+
+We currently solve this in Feature flags with a feature for syncing flags between projects, based on the flag key. Feature flags make this uniquely easy thanks to the feature flag key being an effective organization-wide identifier. No other entity – dashboards, insights, etc. – has this advantage though.
+
+Because feature flags use cohorts and actions in their definitions, these are also synced in the current solution –but only based on the cohort/action name, so if the name is changed in one environment, the link is silently broken.
+
+
+Previous context on this topic: [the December 2022 "Feature flag environments" RFC](https://github.com/PostHog/product-internal/blob/main/requests-for-comments/2022-12-08-feature-flag-environments.md#221115---rfc-feature-flag-environments)
+
+## Expectations
+
+### Isolated
+
+The key feature is separation of all of *data*:
+
+- Events
+- People
+- Groups
+- Recordings
+
+Pipeline setup should probably be separate too:
+
+- Apps
+- Batch exports
+
+Domain-specific configuration is also going to differ between environments likely:
+
+- Toolbar authorized URLs
+
+### Shared
+
+Sharing of these entities is commonly requested:
+
+- Dashboards
+- Insights
+- Actions
+- Cohorts
+- Feature flags
+
+Not mentioned typically, but logically also shared should be:
+
+- Event and property definitions
+- Group types
+- Annotations
+- Playlists
+- Experiments
+- Surveys
+- Early access features
+- Project settings
+
+The needs are unclear for:
+
+- Data warehouse tables
+
+## Rejected solution: Cross-project copying
+
+This is what some users have asked for verbatim. What it would probably mean is a feature flags-style "Projects" view on the item level, allowing an item to be copied/synced across projects. The reliability of syncing is key though, and the problem mentioned before is that there's no inherent cross-project ID for this. Just basing on the entity name would be extremely fragile – there needs to be an underlying link better than that.
+
+This means we'd need to maintain such underlying links (e.g. a cross-project ID, adjacent to the primary key ID) for all entities we'd like to work cross-project. In one way, this could be nice if you'd like to develop, say, a dashboard in the test environment and only copy it to the production environment when final. Divergence becomes easy. But there would be an act of manual syncing, making the experience cumbersome for users at scale. This would work, but it wouldn't be turnkey.
+
+[insert metric for how often flags differ within org]
+
+## Proposed solution: Environments
```

Comment:
It is not really clear to me what the difference is between multiple projects vs multiple environments. As in, do we just add another level of hierarchy, or would "environments" support different functionality compared to projects?

You have slightly been referring to it in you response to Neil below, but I want to confirm my assumption is correct.

Each environment contains the same "Dashboards, Insights, Actions, Cohorts, Feature flags, etc." because they are being set once on the project level and then available in any environment. Is that correct?